dimod/binary/pybqm.py

```diff
@@ -0,0 +1,336 @@
+        except KeyError:
+            raise ValueError(f"unknown variable {v!r}") from None
+        for u, bias in Nv.items():
+            if u != v:
+                yield u, bias
+
+    def iter_quadratic(self) -> Iterator[Tuple[Variable, Variable, Any]]:
+        seen = set()
+        for u, Nu in self._adj.items():
+            seen.add(u)
+            for v, bias in Nu.items():
+                if v not in seen:
+                    yield u, v, bias
+
+    def num_variables(self) -> int:
+        return len(self._adj)
+
+    def num_interactions(self) -> int:
+        n = sum(map(len, self._adj.values()), 0)
+        n -= self.num_variables()  # subtract the self-loops
+        return n // 2
+
+    def reduce_linear(self, function: Callable,
+                      initializer: Optional[Any] = None) -> Any:
+        gen = (self.get_linear(v) for v in self.variables)
+        if initializer is None:
+            return reduce(function, gen)
+        else:
+            return reduce(function, gen, initializer)
+
+    def reduce_neighborhood(self, v: Variable, function: Callable,
+                            initializer: Optional[Any] = None) -> Any:
+        gen = (b for _, b in self.iter_neighborhood(v))
+        if initializer is None:
+            return reduce(function, gen)
+        else:
+            return reduce(function, gen, initializer)
+
+    def reduce_quadratic(self, function: Callable,
+                         initializer: Optional[Any] = None) -> Any:
+        gen = (b for _, _, b in self.iter_quadratic())
+        if initializer is None:
+            return reduce(function, gen)
+        else:
+            return reduce(function, gen, initializer)
+
+    def remove_interaction(self, u: Variable, v: Variable):
+        if u == v:
+            raise ValueError(f"{u!r} cannot have an interaction with itself")
+        try:
+            self._adj[u].pop(v)
+        except KeyError:
+            raise ValueError(f"{u!r} and {v!r} have no interaction") from None
+
+        self._adj[v].pop(u)
+
+    def relabel_variables(self, mapping: Mapping[Variable, Variable]):
+        adj = self._adj
+
+        for submap in iter_safe_relabels(mapping, self.variables):
+            for old, new in submap.items():
+                if old == new:
+                    continue
+
+                # replace the linear bias
+                adj[new] = {new: adj[old].pop(old)}
+
+                # copy the quadratic biases
+                for v in adj[old]:
+                    adj[new][v] = adj[v][new] = adj[v].pop(old)
+
+                # remove the old adj for old
+                del adj[old]
+
+    def relabel_variables_as_integers(self) -> Mapping[int, Variable]:
+        mapping = dict((v, i) for i, v in enumerate(self.variables) if i != v)
+        self.relabel_variables(mapping)
+        return dict((i, v) for v, i in mapping.items())
+
+    def remove_variable(self, v: Optional[Variable] = None) -> Variable:
```

Comment:
The linear / quadratic of a variable got absorbed to the offset in case of spin based bqm. Do we want to account for that when removing a variable ? Like remove the contribution it made to the offset ?

Reply:
That's what `fix_variable` does, `remove_variable` is just a raw removal.